akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala

```diff
@@ -0,0 +1,126 @@
+                                         val reseedingThreshold: Long     = CounterRNGConstants.RESEEDING_THRESHOLD,
+                                         val reseedingDeadline:  Long     = CounterRNGConstants.RESEEDING_DEADLINE,
+                                         val reseedingTimeout:   Duration = CounterRNGConstants.RESEEDING_TIMEOUT) extends Random {
+  import CounterRNGConstants._
+
+  private val entropySource = new SecureRandom
+
+  // mutable state below, concurrent accesses need synchronized or lock
+  private val counter: Array[Byte] = Array.fill[Byte](COUNTER_SIZE_BYTES)(0)
+  private var index: Int = 0
+  private var currentBlock: Array[Byte] = null
+  private var reseedFuture: Future[Array[Byte]] = null
+  private var bitsSinceSeeding: Long = 0
+
+  private val cipher = Cipher.getInstance("AES/CTR/NoPadding")
+  private val ivArr = Array.fill[Byte](COUNTER_SIZE_BYTES)(0)
+  ivArr(0) = (ivArr(0) + 1.toByte).toByte
+  private val ivSpec = new IvParameterSpec(ivArr)
+  cipher.init(Cipher.ENCRYPT_MODE, new this.AESKey(seed), ivSpec)
+
+  @Override
+  override protected def next(bits: Int): Int = synchronized {
+    // random result generation phase - if there is not enough bits in counter variable
+    // we generate some more with AES/CTR
+    bitsSinceSeeding += bits
+    if (currentBlock == null || currentBlock.length - index < 4) {
+      try {
+        currentBlock = cipher.doFinal(counter)
+        index = 0
+      } catch {
+        case ex: Exception ⇒
+          // Generally Cipher.doFinal() from nextBlock may throw various exceptions.
+          // However this should never happen.  If initialisation succeeds without exceptions
+          // we should be able to proceed indefinitely without exceptions.
+          throw new IllegalStateException("Failed creating next random block.", ex)
+      }
+    }
+
+    // now, enough bits in counter, generate pseudo-random result
+    val result = (BITWISE_BYTE_TO_INT & currentBlock(index + 3)) |
+      ((BITWISE_BYTE_TO_INT & currentBlock(index + 2)) << 8) |
+      ((BITWISE_BYTE_TO_INT & currentBlock(index + 1)) << 16) |
+      ((BITWISE_BYTE_TO_INT & currentBlock(index)) << 24)
+
+    // re-seeding phase
+    // first, we check if reseedingThreshold is exceeded to see if new entropy is required
+    // we can still proceed without it, but we should ask for it
+    if (bitsSinceSeeding > reseedingThreshold) {
+      if (reseedFuture == null) {
+        // ask for a seed and process async on a separate thread
+        reseedFuture = Future { entropySource.generateSeed(32) }
+      }
+      // check if reseedingDeadline is exceeded - in that case we cannot proceed, as that would be insecure
+      // we need to block on the future to wait for entropy
+      if (bitsSinceSeeding > reseedingDeadline) {
+        try {
+          Await.ready(reseedFuture, reseedingTimeout)
```

Comment:
Default time-out is 5 minutes. Something would need to be seriously wrong to exceed